_posts/2020-05-07-removing-capybara-webkit.md

```diff
@@ -0,0 +1,133 @@
+- It's difficult to update the webkit engine that's run in the background so you can't have all the new features you would have on the actual browsers
+
+## Process
+
+We will devide this change into 4 different steps:
+1. Replace `capybara-webkit`
+1. Make sure specs passes locally
+2. Make sure specs passes inside the Docker container
+3. Make sure specs passes in CircleCI
+
+## Replacing
+
+The first thing we did was to remove the gem from the Gemfile and then added the `webdrivers` gem (which is already included in any new Rails project).
+
+```diff
+- gem "capybara-webkit"
++ gem 'webdrivers'
+```
+
+This project was old so it also had an old capybara version that won't automatically register the browser drivers that we want so we had to register the headless browsers' drivers on our capybara config:
+
+```ruby
+# from https://github.com/teamcapybara/capybara/blob/c7c22789b7aaf6c1515bf6e68f00bfe074cf8fc1/lib/capybara/registrations/drivers.rb
+
+Capybara.register_driver :headless_firefox do |app|
+  Capybara::Selenium::Driver.load_selenium
+  browser_options = ::Selenium::WebDriver::Firefox::Options.new
+  browser_options.args << '-headless'
+  Capybara::Selenium::Driver.new(app, browser: :firefox, options: browser_options)
+end
+
+Capybara.register_driver :headless_chrome do |app|
+  Capybara::Selenium::Driver.load_selenium
+  browser_options = ::Selenium::WebDriver::Chrome::Options.new.tap do |opts|
+    opts.args << '--headless'
+    opts.args << '--disable-gpu' if Gem.win_platform?
+    # Workaround https://bugs.chromium.org/p/chromedriver/issues/detail?id=2650&q=load&sort=-id&colspec=ID%20Status%20Pri%20Owner%20Summary
+    opts.args << '--disable-site-isolation-trials'
+  end
+  Capybara::Selenium::Driver.new(app, browser: :chrome, options: browser_options)
+end
+```
+> You'll need Chrome or Firefox versions that supports headless mode
+
+And now we tell capybara to use one of those drivers for all the tests that require javascript:
+
+```ruby
+Capybara.javascript_driver = :headless_chrome # or :headless_firefox
+```
+
+> We had an extra step here because a lot of specs had the drivers specified on the actual test `describe` block like this: `describe "Do something", js: true, driver: :webkit do`. Instead of replacing the driver there, we just removed that option on all the tests so it uses the base config we set above.
+
+And lastly, before you run your test, make you you have no references to capybara-webkit in your code (search for `capybara-webkit`, `Capybara::Webkit` and `:webkit` strings).
+
+# Fixing Specs
+
+We first make sure tests are working locally so we are sure we don't have errors due to a misconfigured container.
+
+This application uses an old version of [Chosen](https://harvesthq.github.io/chosen/) to customize the `select` tags. The test suite includes a helper method so the driver can select options from that custom select and that helper method started failing.
```

Comment:
Just curious if this was related to the `capybara-webkit` change?

Reply:
This errors were only present on that branch after replacing the gem, and I think it makes sense since capybara-webkit was using an old deprecated QtWebView (which I think was bugged since I had another error where it wasn't actually hiding the original SELECT tag when replacing it with chosen and an expectation was succeeding when it should have failed)